src/dune_rules/fetch_rules.ml

```diff
@@ -0,0 +1,275 @@
+    type target = Path.Build.t
+
+    module Spec = Spec
+
+    let v = { Spec.target; checksum; url; kind }
+  end
+  in
+  Action.Extension (module M)
+;;
+
+let url_of_source (source : Dune_pkg.Source.t) checksum =
+  match source with
+  | Fetch { url; checksum = Some ((_, checksum') as checksum_with_loc) } ->
+    if Checksum.equal checksum checksum' then Some (url, checksum_with_loc) else None
+  | _ -> None
+;;
+
+let url_of_checksum ctx checksum =
+  let+ lock_dir = Lock_dir.get ctx in
+  (* CR-rgrinberg: we should probably index the lock dir by the checksums *)
+  Package.Name.Map.values lock_dir.packages
+  |> List.find_map ~f:(fun (pkg : Pkg.t) ->
+    let source =
+      match pkg.info.source with
+      | None -> None
+      | Some source -> url_of_source source checksum
+    in
+    let extra_sources =
+      List.find_map pkg.info.extra_sources ~f:(fun (_, source) ->
+        url_of_source source checksum)
+    in
+    Option.first_some source extra_sources)
+;;
+
+let gen_rules_checksum url checksum =
+  (* CR-rgrinberg: it's possible to share the downloading step between the
+     directory and file actions. Though it's unlikely to be of any use in real
+     world situations. *)
+  let rule =
+    let info = Rule.Info.of_loc_opt (Some (fst url)) in
+    fun { Action_builder.With_targets.build; targets } ->
+      Rules.Produce.rule (Rule.make ~info ~targets build)
+  in
+  let make_target = make_target (snd checksum) in
+  let action ~target ~kind =
+    action ~url ~checksum:(Some checksum) ~target ~kind
+    |> Action.Full.make
+    |> Action_builder.With_targets.return
+  in
+  let dir_rule =
+    let target = make_target ~kind:`Directory in
+    action ~target ~kind:`Directory
+    |> Action_builder.With_targets.add_directories ~directory_targets:[ target ]
+  in
+  let file_rule =
+    let target = make_target ~kind:`File in
+    action ~target ~kind:`File |> Action_builder.With_targets.add ~file_targets:[ target ]
+  in
+  let open Memo.O in
+  let+ () = rule dir_rule
+  and+ () = rule file_rule in
+  ()
+;;
+
+let gen_rules ctx ~dir ~components =
+  match components with
+  | [] ->
+    Memo.return Rules.empty
+    |> Gen_rules.make
+         ~build_dir_only_sub_dirs:
+           (Gen_rules.Build_only_sub_dirs.singleton ~dir Subdir_set.all)
+    |> Memo.return
+  | [ checksum ] ->
+    let checksum = Dune_pkg.Checksum.parse_string_exn (Loc.none, checksum) in
+    let directory_targets =
+      let target_dir = make_target ~kind:`Directory checksum in
+      Path.Build.Map.singleton target_dir Loc.none
+    in
+    let rules =
+      Rules.collect_unit (fun () ->
+        url_of_checksum ctx checksum
+        >>= function
+        | None ->
+          User_error.raise
+            [ Pp.textf "unknown checksum %s" (Checksum.to_string checksum) ]
+        | Some (url, checksum) -> gen_rules_checksum url checksum)
+    in
+    Gen_rules.make rules ~directory_targets |> Memo.return
+  | _ -> Memo.return Gen_rules.no_rules
```

Comment:
What is `compontents` here? Is silently no-oping in that case the right thing to do?

Reply:
components would be any other build path `_build/_fetch/` that doesn't match the pattern above. So it's correct to return no rules for these unknown paths. There's not much of a real difference though. We load the rules in a directory to build a particular target in it, so we could either:

1. Get an error loading some directory rules
2. Depending on that target and then dune telling us the target doesn't exist.